---
title: "Exercise for Lecture 6 - more on differential expression with the limma package"
output: html_document
---

The purpose of this exercise is to understand a few more details of a standard 'limma' differential expression (DE) analysis.  In particular, we will explore: 
  1. the combination of design matrices and contrast matrices to answer DE questions-of-interest
  2. some of the preprocessing steps (and the concepts leading to them) for Affymetrix microarray data.
  
If you need additional resources to understand this exercise or the methods behind it, it is strongly encourage to read both the [http://www.statsci.org/smyth/pubs/ebayes.pdf](limma paper) (as given in class) and the [https://www.bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf](limma user's guide).


```{r warning=FALSE, message=FALSE}
.libPaths("C:/Users/statzerc/Desktop/%HOMESHARE%/R3UserLibs") #set the library
library("limma")
#source("https://bioconductor.org/biocLite.R")
#biocLite("affy")
library("affy")
#biocLite("preprocessCore")
library("preprocessCore")
unzip("affy_estrogen.zip")
ddir <- "affy_estrogen"
dir(ddir)
```

It is generally good practice to store the details of an experiment (e.g., the set of samples) in a machine-readable table, like the provided ``targets.txt'' file; this is known as **metadata**.  Have a look at this file in a text editor or a spreadsheet to see what kind of information is typically described.  The following code reads in this metadata file, reads in the Affymetrix data and processes it with a popular method called RMA (robust multichip analysis); some of the details of RMA are reverse-engineered below.

```{r warning=FALSE, message=FALSE}
# preprocess affymetrix data
targets <- readTargets("targets.txt", path=ddir)
targets

abatch <- ReadAffy(filenames=targets$filename,
                   celfile.path=ddir)
eset <- rma(abatch)  # bg correct, normalize, summarize
```


It is always good practice to look at overall summaries of a large dataset, such as a multidimensional scaling (MDS) plot to get an idea of the relations between samples.  In this case, "distances on the plot approximate the typical log2 fold changes" (?plotMDS):

```{r}
plotMDS( exprs(eset) )  # MDS plot
```

In order to run the standard limma pipeline for differential expression, we need a design matrix and optionally, a contrast matrix.  In the code below, the metadata is encoded into a factor variable that is used for creating the design matrix.  It is suggested to look at and understand the design matrix before proceeding.

```{r}
# do the limma modeling
f <- paste(targets$estrogen,targets$time.h,sep="")
f <- factor(f)

# create design matrix
design <- model.matrix(~0+f)
colnames(design) <- levels(f)
design
```



From the design matrix, we can now **fit** the linear model (for each gene):

```{r}
fit <- lmFit(eset, design)
```

To ask questions about the parameters defined in the design matrix, we can now define a **contrast** matrix, which can be constructed by hand or using the makeContrasts() accessory function.  Again, it is suggested to study this matrix and make sure you understand what it is doing before proceeding.

```{r}
cont.matrix <- makeContrasts(E10="present10-absent10",
                             E48="present48-absent48",
                             Time="absent48-absent10",levels=design)
cont.matrix
```

Now, the contrasts can be fit and the moderation of the variance parameters (as discussed in lectures) can be performed:

```{r}
fit2  <- contrasts.fit(fit, cont.matrix)
fit2  <- eBayes(fit2)
fit2
class(fit2)
names(fit2)
```

At this point, a lot of elements have now been added to the ``fit2'' object and it would again be worth studying the details.  See if you can understand what the different components are and if you need more details, type ?"MArrayLM-class" to see more information.

Next, we wish to compute the differential expression statistics, such as moderated-t and F statistics and perhaps some P-values.  The topTable() function has many facilities for this:

```{r}
topTable(fit2,coef=1)
topTable(fit2,coef=2)
```

Here, it is worth understanding exactly what "coef=1" or "coef=2" are testing before proceeding.  

It is also recommended to look at the data that goes into the statistical test in the first place.  For example, a simple barplot() for one gene's expression values:

```{r}
barplot( exprs(eset)["39642_at",], las=2, cex.names=.7 )  # top gene (for coef 1)
```

#### Question 1. From the matrix of summarized Affymetrix data that went into the limma pipeline in the first place -- exprs(eset) -- manually calculate the logFC and AveExpr for one of the top differentially expressed genes.
Important side note: if you experiment is sufficiently simple (e.g., a 1-way design), you may not need a contrast matrix at all; the differential expression of interest can be defined through column(s) of the design matrix (see discussion in the lecture notes).


1.) Start at exprs(eset)
2.) calculate logFC and AceExpr for 39642_at
  - define this by the columns of the design matrix
  
What limma does is it assumes the values are all in the log space. Thus, logFC is substracting the one group average for a gene from the other. I am calculating the value for the first coefficient, since this is the value for which this gene set ID was ranked top: 39642_at
```{r}
GeneX <- exprs(eset)["39642_at",]
mean(GeneX[c(3,4)])-mean(GeneX[c(1,2)])
```
This value is the same as can be taken from the TopTable output under "logFC".

```{r}
#The AveExpr value is generated by taking the mean of all expression values for that particular gene set ID across all conditions (8 conditions here)
mean(GeneX)
```
This value is the same as can be taken from the TopTable output under AveExpr.


#### Question 2. Create a Venn diagram summarizing the numbers of  differential genes for the 3 contrasts defined above 'E10', 'E48' and 'Time'.  Hint: use the decideTests() and vennDiagram() functions in limma (you can use default settings).  If you need more help using these functions, try: ?decideTests, ?vennDiagram

Plan:
1.) Get the list of genes that are significantly diff. expressed for each contrast: decideTest(fit2) --> output: another matrix
2.) Plot using vennDiagram on the matrix.

```{r}
sign_reg <- decideTests(fit2)
vennDiagram(sign_reg,circle.col = c("red", "green","blue"))
```
 
#### Question 3. Try to reproduce the above limma modeling using an **alternative design matrix** (remember, it's the combination of a design matrix and a contrast matrix).  Hint: use the experimental factors defined in the 'targets' data frame.

1.) generate a new design and contrast matrix - however, keep the information the same! just use other parameters to describe the experiment
2.) run it again.

#####Explanation of the 'design matrix' in my own words:
The design matrix is basically detailing the influence of the parameters on the experiment, For all combinations of parameters that are specified we test a linear model on the data. Specifying a ZERO (model.matrix(~0 + f)) dictates that there is no intercept in the design matrix. To my understanding one can choose to use an intercept or not since even though the value of the fitting parameters change but the output should not.

#####Contrast function explained in my own words:
If you want to compare two groups and neither of them is the standard you have to use the contrast. If I modify the design matrix and introduce an intercept I have to adjust the contrasts.

#####(Too) Simplified solution to problem:
I will try to adapt this analysis later include all comparisons of the original design matrix but in a different form (as I think the excercise is meant).

```{r}
#I tried to recreate the design matrix with an intercept to make an equivalent analysis. However, I did not succeed. What worked however is an aggregated/ simplified analysis to only compare present vs. absent of the estrogen. This is what I did in the code below.

f <- paste(targets$estrogen,sep="") #I left the time dependency away -> different design matrix BUT also different outcome / comparison
f <- factor(f)
design <- model.matrix(~0 +f)
colnames(design) <- levels(f)
design
```

```{r}
fit <- lmFit(eset, design)
cont.matrix <- makeContrasts(ABS="present-absent",levels=design)
cont.matrix

fit2  <- contrasts.fit(fit, cont.matrix)
fit2  <- eBayes(fit2)
fit2
class(fit2)
names(fit2)

topTable(fit2,coef=1) #Only one coef now

sign_reg <- decideTests(fit2)
```

#### Note: Submit both an Rmarkdown/markdown file as well as a compiled HTML file to your private github repository.